File: 84. Largest Rectangle in Histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(0)        
        stack=[]
        res=0
        for i,h in enumerate(heights):
            start=i
            while stack and stack[-1][1]>h:
                index,height=stack.pop()
                area=height*(i-index)
                res=max(res,area)
                start=index
            stack.append([start,h])
        return res
            

File: 84. Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        vector <vector<int>> stack;
        vector <int> vec;
        int index, val,start=0,res=0, temp=0;
        heights.push_back(0);
        for (int i=0;i<heights.size();i++){
            start=i;
            while (stack.size()>0 && stack.back()[1]>heights[i]){
                index=stack.back()[0];
                val=stack.back()[1];
                res=max(res,val*(i-index));
                start=index;
                stack.pop_back();
            }
            stack.push_back(vector <int> {start, heights[i]});

           
        }
        return res;      
    }
};
